lib/sqs.js

```diff
@@ -0,0 +1,145 @@
+'use strict';
+
+const assert = require('assert');
+const AWS    = require('aws-sdk');
+
+
+module.exports = class SQS {
+  constructor(config) {
+    this.accessKeyId     = config.accessKeyId;
+    this.secretAccessKey = config.secretAccessKey;
+    this.region          = config.region;
+    this.queueName       = config.queueName;
+  }
+
+  post(messages) {
+    return this.sqsClient()
+      .then(sqsClient => {
+        const entries = messages.map((message, index) => {
+          const entry = {
+            // Result may be unordered, that's why we need to provide an ID
+            // for each message in this batch.
+            Id:          index.toString(),
+            MessageBody: message.body
+          };
+
+          // This is the recommended way to check if a queue is FIFO or not.
+          if (this.queueName.endsWith('.fifo'))
+            entry.MessageGroupId = this.queueName;
+
+          return entry;
+        });
+
+        const sqsPayload = {
+          QueueUrl: this.queueURL,
+          Entries:  entries
+        };
+
+        return sqsClient.sendMessageBatch(sqsPayload).promise();
+      })
+      .then(response => {
+        const successful = response.Successful;
+        const failed     = response.Failed;
+
+        if (failed.length) {
+          const errors = failed.map(error => `${error.Code}: ${error.Message}`);
+          throw new Error(errors.join(', '));
+        }
+
+        // Reconstruct array of job IDs based on the batch ID for this request.
+        const orderedJobIDs = messages.map(function(message, index) {
+          const responseResult = successful.find(msg => msg.Id === index.toString());
+          assert(responseResult);
```

Comment:
> could this ever fail?

If I knew, I'd put a try/catch. I think assertions in the code are about expressing assumptions. Here I'm expressing the assumption that for every message we sent, there must be a result in the response, matched by the request ID (temporary ID we pass to the batch operation).

If I remove the assertion, and some assumption I'm making doesn't hold (all result objects have a string `Id`, there's always a result for every message sent), we could inadvertently end up with nulls in the mapped array. (In this case, because we traverse the object, we'd get another error: `TypeError: Cannot read property 'MessageId' of null`.)

So, removing the assertion in this specific case has a similar effect about the behavior I want: throw an error, no silent bugs.